### PR TITLE
feat(dart): Add support for Dart

### LINKED ISSUE #3301: Support pub package manager
Add support for parsing `pubspec.yml` files. Description is in https://github.com/renovatebot/renovate/blob/master/lib/manager/pub/readme.md

Waiting on #3300 (datasource) first

### PR DESCRIPTION
**Pull request by zharinov**
It works for most of Dart pubfiles (except invalid YAML). Later maybe we should implement own "weaker" parser.

Closes #3301

### CODE REVIEW

**Comment by zharinov** [comment]
@rarkins Please take a look at snapshots mismatch, not sure if I can update it right

File: lib/manager/pub/index.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    const doc = yaml.safeLoad(content);
+    return [
+      ...parseDeps(doc.dependencies, {
+        depType: 'dependencies',
+      }),
+      ...parseDeps(doc.dev_dependencies, {
+        depType: 'dev_dependencies',
+      }),
+    ];
+  } catch (e) {
```
Comment (rarkins) [inline lib/manager/pub/index.js]: Let's log here

File: lib/manager/pub/index.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    return null;
+  }
+}
+
+async function extractAllPackageFiles(config, packageFiles) {
```
Comment (rarkins) [inline lib/manager/pub/index.js]: I don't think we need this function - it should function without it

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@zharinov snapshots updated/tests passing

Can you describe more about your update algorithm? Is it based on any of our existing managers, or something new?

**Comment by zharinov** [comment]
I assume it uses npm algorithm since we're using its `versionSchema`

**Comment by rarkins** [comment]
I meant your approach to iterating line by line and the various logic inside the for loop. I'm not sure we've done it that way before and how much testing it needs

**Comment by zharinov** [comment]
> I meant your approach to iterating line by line and the various logic inside the for loop. I'm not sure we've done it that way before and how much testing it needs
> 
Yaml parser doesn’t support metadata, so i’ve done this ad-hoc way. At the other hand, parse-update-dump cycle produces lots of garbage in diffs. As I can see, some other yaml-based managers doing this too, though format differs.

**Comment by rarkins** [comment]
Can you explain the metadata problem more?

**Comment by zharinov** [comment]
Line and column numbers of parsed pubspec files can't collected by the library. For example, in maven we use `fileReplacePosition` field for this stuff, so it can be updated relatively easy.

**Comment by rarkins** [comment]
@zharinov is this ready for merge, in your opinion?

**Comment by zharinov** [comment]
I think yes, at least just to gather more feedback

**Pull request merged**